loader/src/main.rs

```diff
@@ -643,8 +643,10 @@ pub fn read_initial_config(cfg: &mut BootConfig) {
 /// Copy program data from the SPI flash into newly-allocated RAM
 /// located at the end of memory space.
 fn copy_processes(cfg: &mut BootConfig) {
+    let mut _pid = 1;
     for tag in cfg.args.iter() {
         if tag.name == u32::from_le_bytes(*b"IniE") {
+            _pid += 1;
             let mut page_addr: usize = 0;
             let mut previous_addr: usize = 0;
             let mut top = core::ptr::null_mut::<u8>();
```

Comment:
Just a question...this works because the idea is the copy starts at address 0, due to the way the memory space is laid out, correct?

Reply:
(referring to top being core::ptr::null_mut with no other obvious initialization before stuff is added to it)

Reply:
The `_pid` is just for debugging. It's normally ignored, which is why it's prefixed by `_`. We assign PIDs in the order in which they were linked, which is why this trick works.

Reply:
`top` needs to be initialized to *something*. It's initialized in part 2, when the bulk of copies happens.

Part 1 is responsible for finishing up a previous copy before moving onto a new section. That part will only be run when moving from one section to another, so in practice `top` will always be initialized to something.